When starting SolarGraph, try to respect project-specific installation

### Check for existing issues

- [X] Completed

### Is your feature request related to a problem?

If a user has multiple versions of the SolarGraph language server, Zed should use the one that's most appropriate for the current project.

### Describe the solution you'd like

From this [comment](https://github.com/zed-industries/feedback/issues/113#issuecomment-1322651896) by @attilagyorffy

I believe Zed should ideally follow the [Convention Over Configuration paradigm](https://en.wikipedia.org/wiki/Convention_over_configuration) and be smart about the above without the need of any configuration necessary. I believe a sensible priority (from higher to lower) would be the following:

* Use the solargraph bin stub if present (`./bin/solargraph`)
* Use the bundled version of solargraph if specified by the bundler context (see method described above)
* Use the matching solargraph in the `$PATH` (which should match the version of Ruby -> This is the current implementation if I understand correctly)
* Fail with a helpful error message, and with a link to a publicly available piece that helps setting things up

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

hey guys. I think I may have the same problem, but wanted to add some details.
Readme says:

> Zed currently doesn't install Solargraph automatically. To use Solargraph, you need to install the gem. Zed just looks for an executable called `solargraph` on your `PATH`

But i don't think this is true. I have installed `solargraph` with my project's Gemfile, not globally. The `solargraph` bin is in `PATH` but Zed still fails to find it, because it looks in wrong place:

```log
2024-02-22T14:28:08+01:00 [INFO] starting language server "solargraph", path: "/Users/i/project", id: 3
2024-02-22T14:28:08+01:00 [INFO] starting language server. binary path: "solargraph", working directory: "/Users/i/project", args: ["stdio"]
2024-02-22T14:28:08+01:00 [ERROR] failed to start language server "solargraph": No such file or directory (os error 2)
2024-02-22T14:28:08+01:00 [ERROR] server stderr: Some("")
2024-02-22T14:28:08+01:00 [INFO] Validation check requested for "solargraph" but it cannot be reinstalled
```

```sh
$ which solargraph
/Users/i/.rvm/gems/ruby-3.2.1@project/bin/solargraph
```

Another thing to mention: I see that `settings.json` allows to set a path to lsp binary:

```json
  "elixir": {
    // Change the LSP zed uses for elixir.
    // Note that changing this setting requires a restart of Zed
    // to take effect.
    //
    // May take 3 values:
    //  1. Use the standard ElixirLS, this is the default
    //         "lsp": "elixir_ls"
    //  2. Use the experimental NextLs
    //         "lsp": "next_ls",
    //  3. Use a language server installed locally on your machine:
    //         "lsp": {
    //           "local": {
    //             "path": "~/next-ls/bin/start",
    //             "arguments": ["--stdio"]
    //            }
    //          },
    //
    "lsp": "elixir_ls"
  }
```

But I believe this setting does not work for ruby.

--- a/docs/src/languages/ruby.md
+++ b/docs/src/languages/ruby.md
@@ -66,6 +66,20 @@ Solargraph has formatting and diagnostics disabled by default. We can tell Zed t
 }
 ```
 
+To use Solargraph in the context of the bundle, you can use [folder-specific settings](../configuring-zed#settings-files) and specify the absolute path to the [`binstub`](https://bundler.io/v2.5/man/bundle-binstubs.1.html) of Solargraph:
+
+```json
+{
+  "lsp": {
+    "solargraph": {
+      "binary": {
+        "path": "<path_to_your_project>/bin/solargraph"
+      }
+    }
+  }
+}
+```
+
 ### Configuration
 
 Solargraph reads its configuration from a file called `.solargraph.yml` in the root of your project. For more information about this file, see the [Solargraph configuration documentation](https://solargraph.org/guides/configuration).
@@ -114,6 +128,20 @@ Rubocop has unsafe autocorrection disabled by default. We can tell Zed to enable
     "rubocop": {
       "initialization_options": {
         "safeAutocorrect": false
+      }
+    }
+  }
+}
+```
+
+To use Rubocop in the context of the bundle, you can use [folder-specific settings](../configuring-zed#settings-files) and specify the absolute path to the [`binstub`](https://bundler.io/v2.5/man/bundle-binstubs.1.html) of Rubocop:
+
+```json
+{
+  "lsp": {
+    "rubocop": {
+      "binary": {
+        "path": "<path_to_your_project>/bin/rubocop"
       }
     }
   }

--- a/extensions/ruby/src/language_servers/rubocop.rs
+++ b/extensions/ruby/src/language_servers/rubocop.rs
@@ -1,6 +1,9 @@
-use zed_extension_api::{self as zed, settings::LspSettings, Result};
+use zed_extension_api::{self as zed, settings::LspSettings, LanguageServerId, Result};
 
-use crate::RubyLanguageServerCommand;
+pub struct RubocopBinary {
+    pub path: String,
+    pub args: Option<Vec<String>>,
+}
 
 pub struct Rubocop {}
 
@@ -12,33 +15,45 @@ impl Rubocop {
     }
 
     pub fn language_server_command(
-        &self,
+        &mut self,
+        language_server_id: &LanguageServerId,
         worktree: &zed::Worktree,
-    ) -> Result<RubyLanguageServerCommand> {
-        let mut binary = None;
-        let mut args = None;
+    ) -> Result<zed::Command> {
+        let binary = self.language_server_binary(language_server_id, worktree)?;
 
-        if let Some(binary_settings) =
-            LspSettings::for_worktree(Rubocop::LANGUAGE_SERVER_ID, worktree)
-                .ok()
-                .and_then(|lsp_settings| lsp_settings.binary)
-        {
-            if let Some(bin_path) = binary_settings.path {
-                binary = Some(bin_path);
-            }
-            if let Some(bin_args) = binary_settings.arguments {
-                args = Some(bin_args);
-            }
+        Ok(zed::Command {
+            command: binary.path,
+            args: binary.args.unwrap_or_else(|| vec!["--lsp".to_string()]),
+            env: worktree.shell_env(),
+        })
+    }
+
+    fn language_server_binary(
+        &self,
+        _language_server_id: &LanguageServerId,
+        worktree: &zed::Worktree,
+    ) -> Result<RubocopBinary> {
+        let binary_settings = LspSettings::for_worktree("rubocop", worktree)
+            .ok()
+            .and_then(|lsp_settings| lsp_settings.binary);
+        let binary_args = binary_settings
+            .as_ref()
+            .and_then(|binary_settings| binary_settings.arguments.clone());
+
+        if let Some(path) = binary_settings.and_then(|binary_settings| binary_settings.path) {
+            return Ok(RubocopBinary {
+                path,
+                args: binary_args,
+            });
         }
-        let command = if let Some(binary) = binary {
-            binary
-        } else {
-            worktree.which("rubocop").ok_or_else(|| {
-                "rubocop must be installed manually. Install it with `gem install rubocop` or specify the 'binary' path to it via local settings.".to_string()
-            })?
-        };
-        let args = args.unwrap_or_else(|| vec!["--lsp".into()]);
 
-        Ok(RubyLanguageServerCommand { command, args })
+        if let Some(path) = worktree.which("rubocop") {
+            return Ok(RubocopBinary {
+                path,
+                args: binary_args,
+            });
+        }
+
+        Err("rubocop must be installed manually. Install it with `gem install rubocop` or specify the 'binary' path to it via local settings.".to_string())
     }
 }

--- a/extensions/ruby/src/language_servers/rubocop.rs
+++ b/extensions/ruby/src/language_servers/rubocop.rs
@@ -1,4 +1,6 @@
-use zed_extension_api::{self as zed, Result};
+use zed_extension_api::{self as zed, settings::LspSettings, Result};
+
+use crate::RubyLanguageServerCommand;
 
 pub struct Rubocop {}
 
@@ -9,11 +11,34 @@ impl Rubocop {
         Self {}
     }
 
-    pub fn server_script_path(&mut self, worktree: &zed::Worktree) -> Result<String> {
-        let path = worktree.which("rubocop").ok_or_else(|| {
-            "rubocop must be installed manually. Install it with `gem install rubocop` or specify the 'binary' path to it via local settings.".to_string()
-        })?;
+    pub fn language_server_command(
+        &self,
+        worktree: &zed::Worktree,
+    ) -> Result<RubyLanguageServerCommand> {
+        let mut binary = None;
+        let mut args = None;
 
-        Ok(path)
+        if let Some(binary_settings) =
+            LspSettings::for_worktree(Rubocop::LANGUAGE_SERVER_ID, worktree)
+                .ok()
+                .and_then(|lsp_settings| lsp_settings.binary)
+        {
+            if let Some(bin_path) = binary_settings.path {
+                binary = Some(bin_path);
+            }
+            if let Some(bin_args) = binary_settings.arguments {
+                args = Some(bin_args);
+            }
+        }
+        let command = if let Some(binary) = binary {
+            binary
+        } else {
+            worktree.which("rubocop").ok_or_else(|| {
+                "rubocop must be installed manually. Install it with `gem install rubocop` or specify the 'binary' path to it via local settings.".to_string()
+            })?
+        };
+        let args = args.unwrap_or_else(|| vec!["--lsp".into()]);
+
+        Ok(RubyLanguageServerCommand { command, args })
     }
 }

--- a/extensions/ruby/src/language_servers/ruby_lsp.rs
+++ b/extensions/ruby/src/language_servers/ruby_lsp.rs
@@ -1,10 +1,14 @@
 use zed_extension_api::{
+    self as zed,
     lsp::{Completion, CompletionKind, Symbol, SymbolKind},
     settings::LspSettings,
-    CodeLabel, CodeLabelSpan, Result, Worktree,
+    CodeLabel, CodeLabelSpan, LanguageServerId, Result,
 };
 
-use crate::RubyLanguageServerCommand;
+pub struct RubyLspBinary {
+    pub path: String,
+    pub args: Option<Vec<String>>,
+}
 
 pub struct RubyLsp {}
 
@@ -16,34 +20,46 @@ impl RubyLsp {
     }
 
     pub fn language_server_command(
+        &mut self,
+        language_server_id: &LanguageServerId,
+        worktree: &zed::Worktree,
+    ) -> Result<zed::Command> {
+        let binary = self.language_server_binary(language_server_id, worktree)?;
+
+        Ok(zed::Command {
+            command: binary.path,
+            args: binary.args.unwrap_or_default(),
+            env: worktree.shell_env(),
+        })
+    }
+
+    fn language_server_binary(
         &self,
-        worktree: &Worktree,
-    ) -> Result<RubyLanguageServerCommand> {
-        let mut binary = None;
-        let mut args = None;
+        _language_server_id: &LanguageServerId,
+        worktree: &zed::Worktree,
+    ) -> Result<RubyLspBinary> {
+        let binary_settings = LspSettings::for_worktree("ruby-lsp", worktree)
+            .ok()
+            .and_then(|lsp_settings| lsp_settings.binary);
+        let binary_args = binary_settings
+            .as_ref()
+            .and_then(|binary_settings| binary_settings.arguments.clone());
 
-        if let Some(binary_settings) =
-            LspSettings::for_worktree(RubyLsp::LANGUAGE_SERVER_ID, worktree)
-                .ok()
-                .and_then(|lsp_settings| lsp_settings.binary)
-        {
-            if let Some(bin_path) = binary_settings.path {
-                binary = Some(bin_path);
-            }
-            if let Some(bin_args) = binary_settings.arguments {
-                args = Some(bin_args);
-            }
+        if let Some(path) = binary_settings.and_then(|binary_settings| binary_settings.path) {
+            return Ok(RubyLspBinary {
+                path,
+                args: binary_args,
+            });
         }
-        let command = if let Some(binary) = binary {
-            binary
-        } else {
-            worktree
-                .which("ruby-lsp")
-                .ok_or_else(|| "ruby-lsp must be installed manually".to_string())?
-        };
-        let args = args.unwrap_or_else(|| vec![]);
 
-        Ok(RubyLanguageServerCommand { command, args })
+        if let Some(path) = worktree.which("ruby-lsp") {
+            return Ok(RubyLspBinary {
+                path,
+                args: binary_args,
+            });
+        }
+
+        Err("ruby-lsp must be installed manually".to_string())
     }
 
     pub fn label_for_completion(&self, completion: Completion) -> Option<CodeLabel> {

--- a/extensions/ruby/src/language_servers/ruby_lsp.rs
+++ b/extensions/ruby/src/language_servers/ruby_lsp.rs
@@ -1,8 +1,10 @@
-use zed::{
+use zed_extension_api::{
     lsp::{Completion, CompletionKind, Symbol, SymbolKind},
-    CodeLabel, CodeLabelSpan,
+    settings::LspSettings,
+    CodeLabel, CodeLabelSpan, Result, Worktree,
 };
-use zed_extension_api::{self as zed, Result};
+
+use crate::RubyLanguageServerCommand;
 
 pub struct RubyLsp {}
 
@@ -13,13 +15,35 @@ impl RubyLsp {
         Self {}
     }
 
-    pub fn server_script_path(&mut self, worktree: &zed::Worktree) -> Result<String> {
-        let path = worktree.which("ruby-lsp").ok_or_else(|| {
-            "ruby-lsp must be installed manually. Install it with `gem install ruby-lsp`."
-                .to_string()
-        })?;
+    pub fn language_server_command(
+        &self,
+        worktree: &Worktree,
+    ) -> Result<RubyLanguageServerCommand> {
+        let mut binary = None;
+        let mut args = None;
 
-        Ok(path)
+        if let Some(binary_settings) =
+            LspSettings::for_worktree(RubyLsp::LANGUAGE_SERVER_ID, worktree)
+                .ok()
+                .and_then(|lsp_settings| lsp_settings.binary)
+        {
+            if let Some(bin_path) = binary_settings.path {
+                binary = Some(bin_path);
+            }
+            if let Some(bin_args) = binary_settings.arguments {
+                args = Some(bin_args);
+            }
+        }
+        let command = if let Some(binary) = binary {
+            binary
+        } else {
+            worktree
+                .which("ruby-lsp")
+                .ok_or_else(|| "ruby-lsp must be installed manually".to_string())?
+        };
+        let args = args.unwrap_or_else(|| vec![]);
+
+        Ok(RubyLanguageServerCommand { command, args })
     }
 
     pub fn label_for_completion(&self, completion: Completion) -> Option<CodeLabel> {

--- a/extensions/ruby/src/language_servers/ruby_lsp.rs
+++ b/extensions/ruby/src/language_servers/ruby_lsp.rs
@@ -59,7 +59,10 @@ impl RubyLsp {
             });
         }
 
-        Err("ruby-lsp must be installed manually".to_string())
+        Err(
+            "ruby-lsp must be installed manually. Install it with `gem install ruby-lsp`."
+                .to_string(),
+        )
     }
 
     pub fn label_for_completion(&self, completion: Completion) -> Option<CodeLabel> {

--- a/extensions/ruby/src/language_servers/solargraph.rs
+++ b/extensions/ruby/src/language_servers/solargraph.rs
@@ -1,9 +1,12 @@
 use zed::lsp::{Completion, CompletionKind, Symbol, SymbolKind};
 use zed::{CodeLabel, CodeLabelSpan};
 use zed_extension_api::settings::LspSettings;
-use zed_extension_api::{self as zed, Result};
+use zed_extension_api::{self as zed, LanguageServerId, Result};
 
-use crate::RubyLanguageServerCommand;
+pub struct SolargraphBinary {
+    pub path: String,
+    pub args: Option<Vec<String>>,
+}
 
 pub struct Solargraph {}
 
@@ -15,34 +18,46 @@ impl Solargraph {
     }
 
     pub fn language_server_command(
-        &self,
+        &mut self,
+        language_server_id: &LanguageServerId,
         worktree: &zed::Worktree,
-    ) -> Result<RubyLanguageServerCommand> {
-        let mut binary = None;
-        let mut args = None;
+    ) -> Result<zed::Command> {
+        let binary = self.language_server_binary(language_server_id, worktree)?;
 
-        if let Some(binary_settings) =
-            LspSettings::for_worktree(Solargraph::LANGUAGE_SERVER_ID, worktree)
-                .ok()
-                .and_then(|lsp_settings| lsp_settings.binary)
-        {
-            if let Some(bin_path) = binary_settings.path {
-                binary = Some(bin_path);
-            }
-            if let Some(bin_args) = binary_settings.arguments {
-                args = Some(bin_args);
-            }
+        Ok(zed::Command {
+            command: binary.path,
+            args: binary.args.unwrap_or_else(|| vec!["stdio".to_string()]),
+            env: worktree.shell_env(),
+        })
+    }
+
+    fn language_server_binary(
+        &self,
+        _language_server_id: &LanguageServerId,
+        worktree: &zed::Worktree,
+    ) -> Result<SolargraphBinary> {
+        let binary_settings = LspSettings::for_worktree("solargraph", worktree)
+            .ok()
+            .and_then(|lsp_settings| lsp_settings.binary);
+        let binary_args = binary_settings
+            .as_ref()
+            .and_then(|binary_settings| binary_settings.arguments.clone());
+
+        if let Some(path) = binary_settings.and_then(|binary_settings| binary_settings.path) {
+            return Ok(SolargraphBinary {
+                path,
+                args: binary_args,
+            });
         }
-        let command = if let Some(binary) = binary {
-            binary
-        } else {
-            worktree
-                .which("solargraph")
-                .ok_or_else(|| "solargraph must be installed manually".to_string())?
-        };
-        let args = args.unwrap_or_else(|| vec!["stdio".into()]);
 
-        Ok(RubyLanguageServerCommand { command, args })
+        if let Some(path) = worktree.which("solargraph") {
+            return Ok(SolargraphBinary {
+                path,
+                args: binary_args,
+            });
+        }
+
+        Err("solargraph must be installed manually".to_string())
     }
 
     pub fn label_for_completion(&self, completion: Completion) -> Option<CodeLabel> {

--- a/extensions/ruby/src/language_servers/solargraph.rs
+++ b/extensions/ruby/src/language_servers/solargraph.rs
@@ -1,6 +1,9 @@
 use zed::lsp::{Completion, CompletionKind, Symbol, SymbolKind};
 use zed::{CodeLabel, CodeLabelSpan};
+use zed_extension_api::settings::LspSettings;
 use zed_extension_api::{self as zed, Result};
+
+use crate::RubyLanguageServerCommand;
 
 pub struct Solargraph {}
 
@@ -11,12 +14,35 @@ impl Solargraph {
         Self {}
     }
 
-    pub fn server_script_path(&mut self, worktree: &zed::Worktree) -> Result<String> {
-        let path = worktree
-            .which("solargraph")
-            .ok_or_else(|| "solargraph must be installed manually".to_string())?;
+    pub fn language_server_command(
+        &self,
+        worktree: &zed::Worktree,
+    ) -> Result<RubyLanguageServerCommand> {
+        let mut binary = None;
+        let mut args = None;
 
-        Ok(path)
+        if let Some(binary_settings) =
+            LspSettings::for_worktree(Solargraph::LANGUAGE_SERVER_ID, worktree)
+                .ok()
+                .and_then(|lsp_settings| lsp_settings.binary)
+        {
+            if let Some(bin_path) = binary_settings.path {
+                binary = Some(bin_path);
+            }
+            if let Some(bin_args) = binary_settings.arguments {
+                args = Some(bin_args);
+            }
+        }
+        let command = if let Some(binary) = binary {
+            binary
+        } else {
+            worktree
+                .which("solargraph")
+                .ok_or_else(|| "solargraph must be installed manually".to_string())?
+        };
+        let args = args.unwrap_or_else(|| vec!["stdio".into()]);
+
+        Ok(RubyLanguageServerCommand { command, args })
     }
 
     pub fn label_for_completion(&self, completion: Completion) -> Option<CodeLabel> {

--- a/extensions/ruby/src/ruby.rs
+++ b/extensions/ruby/src/ruby.rs
@@ -13,11 +13,6 @@ struct RubyExtension {
     rubocop: Option<Rubocop>,
 }
 
-struct RubyLanguageServerCommand {
-    command: String,
-    args: Vec<String>,
-}
-
 impl zed::Extension for RubyExtension {
     fn new() -> Self {
         Self {
@@ -32,27 +27,21 @@ impl zed::Extension for RubyExtension {
         language_server_id: &LanguageServerId,
         worktree: &zed::Worktree,
     ) -> Result<zed::Command> {
-        let language_server_command = match language_server_id.as_ref() {
-            Solargraph::LANGUAGE_SERVER_ID => self
-                .solargraph
-                .get_or_insert_with(|| Solargraph::new())
-                .language_server_command(worktree)?,
-            RubyLsp::LANGUAGE_SERVER_ID => self
-                .ruby_lsp
-                .get_or_insert_with(|| RubyLsp::new())
-                .language_server_command(worktree)?,
-            Rubocop::LANGUAGE_SERVER_ID => self
-                .rubocop
-                .get_or_insert_with(|| Rubocop::new())
-                .language_server_command(worktree)?,
-            _ => return Err(format!("Unknown language server: {language_server_id}")),
-        };
-
-        Ok(zed::Command {
-            command: language_server_command.command,
-            args: language_server_command.args,
-            env: worktree.shell_env(),
-        })
+        match language_server_id.as_ref() {
+            Solargraph::LANGUAGE_SERVER_ID => {
+                let solargraph = self.solargraph.get_or_insert_with(|| Solargraph::new());
+                solargraph.language_server_command(language_server_id, worktree)
+            }
+            RubyLsp::LANGUAGE_SERVER_ID => {
+                let ruby_lsp = self.ruby_lsp.get_or_insert_with(|| RubyLsp::new());
+                ruby_lsp.language_server_command(language_server_id, worktree)
+            }
+            Rubocop::LANGUAGE_SERVER_ID => {
+                let rubocop = self.rubocop.get_or_insert_with(|| Rubocop::new());
+                rubocop.language_server_command(language_server_id, worktree)
+            }
+            language_server_id => Err(format!("unknown language server: {language_server_id}")),
+        }
     }
 
     fn label_for_symbol(

--- a/extensions/ruby/src/ruby.rs
+++ b/extensions/ruby/src/ruby.rs
@@ -13,6 +13,11 @@ struct RubyExtension {
     rubocop: Option<Rubocop>,
 }
 
+struct RubyLanguageServerCommand {
+    command: String,
+    args: Vec<String>,
+}
+
 impl zed::Extension for RubyExtension {
     fn new() -> Self {
         Self {
@@ -27,36 +32,27 @@ impl zed::Extension for RubyExtension {
         language_server_id: &LanguageServerId,
         worktree: &zed::Worktree,
     ) -> Result<zed::Command> {
-        match language_server_id.as_ref() {
-            Solargraph::LANGUAGE_SERVER_ID => {
-                let solargraph = self.solargraph.get_or_insert_with(|| Solargraph::new());
+        let language_server_command = match language_server_id.as_ref() {
+            Solargraph::LANGUAGE_SERVER_ID => self
+                .solargraph
+                .get_or_insert_with(|| Solargraph::new())
+                .language_server_command(worktree)?,
+            RubyLsp::LANGUAGE_SERVER_ID => self
+                .ruby_lsp
+                .get_or_insert_with(|| RubyLsp::new())
+                .language_server_command(worktree)?,
+            Rubocop::LANGUAGE_SERVER_ID => self
+                .rubocop
+                .get_or_insert_with(|| Rubocop::new())
+                .language_server_command(worktree)?,
+            _ => return Err(format!("Unknown language server: {language_server_id}")),
+        };
 
-                Ok(zed::Command {
-                    command: solargraph.server_script_path(worktree)?,
-                    args: vec!["stdio".into()],
-                    env: worktree.shell_env(),
-                })
-            }
-            RubyLsp::LANGUAGE_SERVER_ID => {
-                let ruby_lsp = self.ruby_lsp.get_or_insert_with(|| RubyLsp::new());
-
-                Ok(zed::Command {
-                    command: ruby_lsp.server_script_path(worktree)?,
-                    args: vec![],
-                    env: worktree.shell_env(),
-                })
-            }
-            Rubocop::LANGUAGE_SERVER_ID => {
-                let rubocop = self.rubocop.get_or_insert_with(|| Rubocop::new());
-
-                Ok(zed::Command {
-                    command: rubocop.server_script_path(worktree)?,
-                    args: vec!["--lsp".into()],
-                    env: worktree.shell_env(),
-                })
-            }
-            language_server_id => Err(format!("unknown language server: {language_server_id}")),
-        }
+        Ok(zed::Command {
+            command: language_server_command.command,
+            args: language_server_command.args,
+            env: worktree.shell_env(),
+        })
     }
 
     fn label_for_symbol(

